include/zephyr/shell/shell.h

```diff
@@ -20,6 +20,12 @@
 #include <zephyr/toolchain.h>
 
 #if defined CONFIG_SHELL_GETOPT
+/*
+ * All shell commands should use the re-entrant APIs provided by the
+ * newlib-compatible getopt APIs. This exposes those and hides the
+ * globals used by the non-reentrant APIs.
+ */
+#define __need_getopt_newlib
```

Comment:
Again, just given that we are trying to move away from supporting newlib, I wonder if this can be done behind the scenes within the build system.

`<getopt.h>` is already non-standard, so I don't believe it would be necessarily wrong to assume that `getopt_r()` and similar reentrant functions are always available within Zephyr.